src/java.base/share/classes/java/nio/Buffer.java

```diff
@@ -747,9 +748,14 @@ final int checkIndex(int i) {                       // package-private
     }
 
     final int checkIndex(int i, int nb) {               // package-private
-        if ((i < 0) || (nb > limit - i))
-            throw new IndexOutOfBoundsException();
-        return i;
+        return Preconditions.checkIndex(i, limit - nb + 1,
+            (x, y) -> {
```

Comment:
I don't know if you've retested the performance but I think you'll need this parameter to be a constant and checkIndex may need to be ForceInline. While we're here, I think we'll have to check the exception message throw by the 1-arg checkIndex to make sure it make sense for buffers.

Reply:
I re-tested with and without ForceInline and see little difference.

Reply:
I was hoping the  3rd parameter to Preconditions.checkIndex would be constant. It looks like it's a capturing lambda ("this" is captured because of the access to the position/limit fields). Hmm, maybe start out simple with a constant, otherwise this change is going to require much wider startup and performance testing.

Reply:
In the original code it just has `throw new IndexOutOfBoundsException();`, so why not make it non-capturing lambda by just returning the same, e.g., `(x, y) -> new IndexOutOfBoundsException()`

Reply:
> Hmm, maybe start out simple with a constant, otherwise this change is going to require much wider startup and performance testing.

Something like?
```
    @ForceInline
    final int checkIndex(int i, int nb) {               // package-private
        return Preconditions.checkIndex(i, limit - nb + 1,
            (x, y) -> new IndexOutOfBoundsException());
    }
```

Reply:
> In the original code it just has `throw new IndexOutOfBoundsException();`, so why not make it non-capturing lambda by just returning the same, e.g., `(x, y) -> new IndexOutOfBoundsException()`

We were trying to improve the message but it looks like it is not worth it.

Reply:
Exception formatter changed to a constant in e4ba71731f806897144a90714a72cdd8516df1d2.